File: 0094-binary-tree-inorder-traversal/solution.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */

void inorder(struct TreeNode* root,int* res,int* ind){
    if(root==NULL){
        return ;
    }

    inorder(root->left,res,ind);
    res[(*ind)++] = root->val;
    inorder(root->right,res,ind);
}

int* inorderTraversal(struct TreeNode* root, int* returnSize) {
    int* res=malloc(sizeof(int)*100);
    //if(!res)return NULL;
    int ind=0;
    inorder(root,res,&ind);
    *returnSize=ind;
    return res;
}

File: 0094-binary-tree-inorder-traversal/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        Stack<TreeNode> stk = new Stack<>();
        TreeNode cur = root;
        while(true){
            while(cur!=null){
                stk.push(cur);
                cur=cur.left;
            }
            if(stk.isEmpty())break;
            cur = stk.pop();
            res.add(cur.val);
            cur=cur.right;
        }
        return res;
    }
}
